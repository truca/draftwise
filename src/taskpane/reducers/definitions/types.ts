export enum DefinitionsTypes {
  INITIALIZE_DEFINITIONS = "INITIALIZE_DEFINITIONS",
  SET_SELECTED_PARAGRAPH_TEXT = "SET_SELECTED_PARAGRAPH_TEXT",
}

export type InitializeAction = {
  type: DefinitionsTypes.INITIALIZE_DEFINITIONS;
  paragraphs: string[];
};

export type SetSelectedParagraphTextAction = {
  type: DefinitionsTypes.SET_SELECTED_PARAGRAPH_TEXT;
  paragraphText: string;
};

export type DefinitionsActions = InitializeAction | SetSelectedParagraphTextAction;

export interface DefinitionsState {
  selectedParagraphText: string;
  paragraphs: string[];
  terms: string[];
  definitionsHash: { [term: string]: string };
}

export type TermsObject = {
  term: string;
  definition: string;
};

export type DefinitionsHash = { [term: string]: string };
