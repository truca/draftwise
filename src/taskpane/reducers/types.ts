export enum DefinitionsTypes {
  INITIALIZE,
}

export type InitializeAction = { type: DefinitionsTypes.INITIALIZE; paragraphs: string[] };

export type DefinitionsActions = InitializeAction;

export interface DefinitionsState {
  paragraphs: string[];
  terms: string[];
  definitionsHash: { [term: string]: string };
}

export type TermsObject = {
  term: string;
  definition: string;
};

export type DefinitionsHash = { [term: string]: string };
