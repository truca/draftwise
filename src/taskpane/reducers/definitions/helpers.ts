import { DefinitionsHash, DefinitionsState, TermsObject } from "./types";

const getTermsFromParagraph = (paragraph: string): TermsObject => {
  const termEndIndex = paragraph.indexOf("”");
  const term = paragraph.substr(1, termEndIndex - 1).toLowerCase();
  const definition = paragraph.substr(termEndIndex + 2);
  return { term, definition };
};

const getHashFromObjects = <ObjectType>(objects: ObjectType[], key: string, valueKey: string): DefinitionsHash => {
  return objects.reduce((accumulator, object) => ({ ...accumulator, [object[key]]: object[valueKey] }), {});
};

const getDefinitionsHashFromParagraphs = (paragraphs: string[]): DefinitionsHash => {
  const paragraphsWithDefinitions = paragraphs.filter((p) => p[0] === "“");
  const termsAndDefinitions = paragraphsWithDefinitions.map(getTermsFromParagraph);

  return getHashFromObjects<TermsObject>(termsAndDefinitions, "term", "definition");
};

export const getDefinitionsStateFromParagraphs = (paragraphs: string[]): Partial<DefinitionsState> => {
  const termsAndDefinitions = getDefinitionsHashFromParagraphs(paragraphs);
  return {
    paragraphs,
    definitionsHash: termsAndDefinitions,
    terms: Object.keys(termsAndDefinitions),
  };
};
