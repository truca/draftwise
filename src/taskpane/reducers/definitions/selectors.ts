import { ReduxState } from "../index";
import { TermsObject } from "./types";

export const getTermsInParagraph =
  (paragraph: string) =>
  (state: ReduxState): string[] => {
    const wordsFromParagraph = paragraph.match(/\w+/g);
    const { terms } = state.definitions;

    return terms.filter((term) => wordsFromParagraph.includes(term));
  };

export const getTermDefinition =
  (term: string) =>
  (state: ReduxState): string | null => {
    const { definitionsHash } = state.definitions;

    return definitionsHash[term];
  };

export const getTerms = (state: ReduxState): string[] => {
  const { terms } = state.definitions;

  return terms;
};

export const getTermsWithDefinitions = (state: ReduxState): TermsObject[] => {
  const terms = getTerms(state);

  return terms.map((term) => {
    const definition = getTermDefinition(term)(state);
    return { term, definition };
  });
};
