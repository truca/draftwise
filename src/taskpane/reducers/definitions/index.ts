import { getDefinitionsStateFromParagraphs } from "./helpers";
import { DefinitionsActions, DefinitionsState, DefinitionsTypes } from "./types";

export const initialState: DefinitionsState = {
  selectedParagraphText: "",
  paragraphs: [],
  terms: [],
  definitionsHash: {},
};

const definitions = (state: DefinitionsState = initialState, action?: DefinitionsActions) => {
  switch (action?.type) {
    case DefinitionsTypes.INITIALIZE_DEFINITIONS:
      // eslint-disable-next-line no-case-declarations
      const definitions = getDefinitionsStateFromParagraphs(action.paragraphs);
      return {
        ...state,
        terms: definitions.terms,
        definitionsHash: definitions.definitionsHash,
        paragraphs: action.paragraphs,
      };
    case DefinitionsTypes.SET_SELECTED_PARAGRAPH_TEXT:
      return { ...state, selectedParagraphText: action.paragraphText };
    default:
      return state;
  }
};

export default definitions;
