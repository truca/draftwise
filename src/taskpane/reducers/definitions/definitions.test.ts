import definitions, { initialState } from "./index";
import {
  getTermsWithDefinitions,
  getTermsWithDefinitionsInParagraph,
  selectedParagraphTextSelector,
} from "./selectors";
import { DefinitionsTypes } from "./types";

const PARAGRAPHS = ["paragraph", "“Expenses” shall include all reasonable attorneys’ fees"];

describe("descriptions", () => {
  describe("reducer", () => {
    it("initial state", () => {
      const state = definitions();
      expect(state).toMatchSnapshot();
    });
    it("initialize definitions", () => {
      const state = definitions(initialState, {
        type: DefinitionsTypes.INITIALIZE_DEFINITIONS,
        paragraphs: PARAGRAPHS,
      });
      expect(state).toMatchSnapshot();
    });
    it("set selected paragraph", () => {
      const state = definitions(initialState, {
        type: DefinitionsTypes.SET_SELECTED_PARAGRAPH_TEXT,
        paragraphText: "paragraph",
      });
      expect(state).toMatchSnapshot();
    });
  });

  describe("selectors", () => {
    it("selectedParagraphText", () => {
      const result = selectedParagraphTextSelector({
        definitions: {
          definitionsHash: {},
          paragraphs: [],
          selectedParagraphText: "paragraph",
          terms: [],
        },
      } as any);
      expect(result).toMatchSnapshot();
    });
    it("getTermsWithDefinitions", () => {
      const result = getTermsWithDefinitions({
        definitions: {
          definitionsHash: {
            expenses: "shall include all reasonable attorneys’ fees",
          },
          paragraphs: ["paragraph", "“Expenses” shall include all reasonable attorneys’ fees"],
          selectedParagraphText: "",
          terms: ["expenses"],
        },
      } as any);
      expect(result).toMatchSnapshot();
    });
    it("getTermsWithDefinitionsInParagraph", () => {
      const result = getTermsWithDefinitionsInParagraph({
        definitions: {
          definitionsHash: {
            expenses: "shall include all reasonable attorneys’ fees",
            fee: "fee",
          },
          paragraphs: ["paragraph", "“Expenses” shall include all reasonable attorneys’ fees"],
          selectedParagraphText: "fee",
          terms: ["expenses", "fee"],
        },
      } as any);
      expect(result).toMatchSnapshot();
    });
  });
});
