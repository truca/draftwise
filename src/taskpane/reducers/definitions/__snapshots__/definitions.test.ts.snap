// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`descriptions reducer initial state 1`] = `
{
  "definitionsHash": {},
  "paragraphs": [],
  "selectedParagraphText": "",
  "terms": [],
}
`;

exports[`descriptions reducer initialize definitions 1`] = `
{
  "definitionsHash": {
    "expenses": "shall include all reasonable attorneys’ fees",
  },
  "paragraphs": [
    "paragraph",
    "“Expenses” shall include all reasonable attorneys’ fees",
  ],
  "selectedParagraphText": "",
  "terms": [
    "expenses",
  ],
}
`;

exports[`descriptions reducer set selected paragraph 1`] = `
{
  "definitionsHash": {},
  "paragraphs": [],
  "selectedParagraphText": "paragraph",
  "terms": [],
}
`;

exports[`descriptions selectors getTermsWithDefinitions 1`] = `
[
  {
    "definition": "shall include all reasonable attorneys’ fees",
    "term": "expenses",
  },
]
`;

exports[`descriptions selectors getTermsWithDefinitionsInParagraph 1`] = `
[
  {
    "definition": "fee",
    "term": "fee",
  },
]
`;

exports[`descriptions selectors selectedParagraphText 1`] = `"paragraph"`;
